node/src/components/linear_chain_sync/operations.rs

```diff
@@ -0,0 +1,288 @@
+    vector_of_peers.shuffle(&mut rng);
+    vector_of_peers
+}
+
+/// Fetch an item. Keeps retrying to fetch until it is successful. Assumes no integrity check is
+/// necessary for the item.
+async fn fetch_retry_forever<T, REv, I>(
+    effect_builder: EffectBuilder<REv>,
+    id: T::Id,
+) -> Result<Box<T>, FetcherError<T, I>>
+where
+    T: Item + 'static,
+    REv: From<FetcherRequest<I, T>> + From<NetworkInfoRequest<I>>,
+    I: Eq + Debug + Clone + Send + 'static,
+{
+    loop {
+        for peer in get_and_shuffle_network_peers(effect_builder).await {
+            trace!(
+                "Attempting to fetch {:?} with id {:?} from {:?}",
+                T::TAG,
+                id,
+                peer
+            );
+            match effect_builder.fetch::<T, I>(id, peer.clone()).await {
+                Ok(FetchedData::FromStorage { item }) => {
+                    trace!(
+                        "Did not get {:?} with id {:?} from {:?}, got from storage instead",
+                        T::TAG,
+                        id,
+                        peer
+                    );
+                    return Ok(item);
+                }
+                Ok(FetchedData::FromPeer { item, .. }) => {
+                    trace!("Fetched {:?} with id {:?} from {:?}", T::TAG, id, peer);
+                    return Ok(item);
+                }
+                Err(FetcherError::AbsentFromPeer { .. }) => {
+                    warn!(
+                        "Fast sync could not fetch {:?} with id {:?} from peer {:?}; trying next peer",
+                        T::TAG,
+                        id,
+                        peer
+                    )
+                }
+                Err(error) => return Err(error),
+            }
+        }
+        tokio::time::sleep(TIMEOUT_DURATION).await
+    }
+}
+
+// fn verify_finality_signatures
+
+/// Fetch a block header from the network by height.
+async fn fetch_block_header_by_height<REv, I>(
+    effect_builder: EffectBuilder<REv>,
+    height: u64,
+    trusted_validator_weights: &BTreeMap<PublicKey, U512>,
+) -> Result<Option<Box<BlockHeaderWithMetadata>>, FetcherError<BlockHeaderWithMetadata, I>>
+where
+    REv: From<FetcherRequest<I, BlockHeaderWithMetadata>> + From<NetworkInfoRequest<I>>,
+    I: Eq + Debug + Clone + Send + 'static,
+{
+    for peer in get_and_shuffle_network_peers(effect_builder).await {
+        match effect_builder
+            .fetch::<BlockHeaderWithMetadata, I>(height, peer.clone())
+            .await
+        {
+            Ok(FetchedData::FromStorage { item }) => return Ok(Some(item)),
+            Ok(FetchedData::FromPeer { item, .. }) => {
+                // TODO: Validate item, ban peer if necessary
+                // Compute the total weight of the validators
+                let _total_weight: U512 = trusted_validator_weights
+                    .iter()
+                    .map(|(_, weight)| *weight)
+                    .sum();
```

Comment:
@afck @fizyk20 @momipsl I started to implement finality signature checking here; should I be doing `1 - ftt` to ensure >66%